Archive/Wiedijk100Theorems/BuffonsNeedle.lean

```diff
@@ -0,0 +1,420 @@
+    intro θ hθ
+    exact mul_nonneg (Real.sin_nonneg_of_mem_Icc hθ) hl.le
+
+  simp_rw [set_integral_toReal_ofReal_nonneg_ae measurableSet_Icc (MeasureTheory.ae_of_all _ this),
+    ← smul_eq_mul, integral_smul_const, smul_eq_mul, mul_comm, mul_eq_mul_left_iff]
+
+  apply Or.inl
+  rw [MeasureTheory.integral_Icc_eq_integral_Ioc, ← intervalIntegral.integral_of_le Real.pi_pos.le,
+    integral_sin, Real.cos_zero, Real.cos_pi]
+
+  ring_nf
+
+/--
+  The integrand in the long case is `min d (θ.sin * l)` and its integrability is necessary for
+  the integral lemmas below.
+-/
+lemma min_const_sin_mul_intervalIntegrable (a b : ℝ) :
+    IntervalIntegrable (fun (θ : ℝ) => min d (θ.sin * l)) ℙ a b := by
+  apply Continuous.intervalIntegrable
+  exact Continuous.min continuous_const (Continuous.mul Real.continuous_sin continuous_const)
+
+/--
+  This equality is useful since `θ.sin` is increasing in `0..π / 2` (but not in `0..π`).
+  Then, `∫ θ in (0)..π / 2, min d (θ.sin * l)` can be split into two adjacent integrals, at the
+  point where `d = θ.sin * l`, which is `θ = (d / l).arcsin`.
+-/
+lemma integral_min_eq_two_mul :
+    ∫ θ in (0)..π, min d (θ.sin * l) = 2 * ∫ θ in (0)..π / 2, min d (θ.sin * l) := by
+  rw [← intervalIntegral.integral_add_adjacent_intervals (b := π / 2) (c := π)]
+  conv => lhs; arg 2; arg 1; intro θ; rw [← neg_neg θ, Real.sin_neg]
+
+  simp_rw [intervalIntegral.integral_comp_neg fun θ => min d (-θ.sin * l), ← Real.sin_add_pi,
+    intervalIntegral.integral_comp_add_right (fun θ => min d (θ.sin * l)), add_left_neg,
+    (by ring : -(π / 2) + π = π / 2), two_mul]
+
+  all_goals exact min_const_sin_mul_intervalIntegrable d l _ _
+
+/--
+  The first of two adjacent integrals in the long case. In the range `(0)..(d / l).arcsin`, we
+  have that `θ.sin * l ≤ d`, and thus the integral is `∫ θ in (0)..(d / l).arcsin, θ.sin * l`.
+-/
+lemma integral_zero_to_arcsin_min :
+    ∫ θ in (0)..(d / l).arcsin, min d (θ.sin * l) = (1 - (1 - (d / l) ^ 2).sqrt) * l := by
+  have : Set.EqOn (fun θ => min d (θ.sin * l)) (Real.sin · * l) (Set.uIcc 0 (d / l).arcsin) := by
+    intro θ ⟨hθ₁, hθ₂⟩
+    have : (d / l).arcsin ≥ 0 := Real.arcsin_nonneg.mpr (div_nonneg hd.le hl.le)
+    simp only [sup_eq_max, inf_eq_min, min_eq_left this, max_eq_right this] at hθ₁ hθ₂
+
+    have hθ_mem : θ ∈ Set.Ioc (-(π / 2)) (π / 2) := by
+      exact ⟨lt_of_lt_of_le (neg_lt_zero.mpr (div_pos Real.pi_pos two_pos)) hθ₁,
+        le_trans hθ₂ (d / l).arcsin_mem_Icc.right⟩
+    simp_rw [min_eq_right ((le_div_iff hl).mp ((Real.le_arcsin_iff_sin_le' hθ_mem).mp hθ₂))]
+
+  rw [intervalIntegral.integral_congr this, intervalIntegral.integral_mul_const, integral_sin,
+    Real.cos_zero, Real.cos_arcsin]
+
+/--
+  The second of two adjacent integrals in the long case. In the range `(d / l).arcsin..(π / 2)`, we
+  have that `d ≤ θ.sin * l`, and thus the integral is `∫ θ in (d / l).arcsin..(π / 2), d`.
+-/
+lemma integral_arcsin_to_pi_div_two_min (h : l ≥ d) :
```

Comment:
We use the spelling `a ≤ b` in general (since all lemmas are writing in this form and it is harder to use whenever we need syntactic equality). Can you change to this convention here and below?

Reply:
Should I do the same for strict equality eg `0 < d`?

Reply:
Yes

Reply:
Although, if you have an assumption of the form 0 < a, then it should be instead phrased as a \neq 0

Reply:
Not sure I understand, `a` is a real number here.

Reply:
I mean to say that the spelling of `d \neq 0` is preferred over `0 < d` if it is an assumption while `0 < d` is preferred over `d \neq 0` if it is the conclusion.

Reply:
I think I'm confused. Those aren't equivalent statements in the reals. So I can't substitute `0 < d` for `d \neq 0` in an assumption, because I also need it to be non-negative.

Reply:
Oh, I'm sorry. I was under the impression that we are talking about `nnreal` rather than `real`. In this case what you have is correct.